common/key_value_parser/key_value_parser.cpp

```diff
@@ -49,6 +49,15 @@ error::Error MakeError(KeyValueParserErrorCode code, const string &msg) {
 
 ExpectedKeyValuesMap ParseKeyValues(const vector<string> &items, char delimiter) {
 	KeyValuesMap ret;
+	error::Error err = AddParseKeyValues(ret, items, delimiter);
+	if (err) {
+		return ExpectedKeyValuesMap(err);
+	} else {
+		return ExpectedKeyValuesMap(ret);
+	}
+}
+
+error::Error AddParseKeyValues(KeyValuesMap &base, const vector<string> &items, char delimiter) {
```

Comment:
S: I would consider returning the map explicitly here, instead of just an error.

Then you can eliminate the `if .. else` above also.

Reply:
That's what I had initially. But it was copying the whole map in and out. How would I avoid that? Taking a reference and then returning it? How would you use a function with such signature then?

Reply:
Why not create the map in the function, and `move` it out of the function afterwards? :thinking: 

```
diff --git a/common/key_value_parser/key_value_parser.cpp b/common/key_value_parser/key_value_parser.cpp
index cc45e14e..8eecc117 100644
--- a/common/key_value_parser/key_value_parser.cpp
+++ b/common/key_value_parser/key_value_parser.cpp
@@ -50,16 +50,7 @@ error::Error MakeError(KeyValueParserErrorCode code, const string &msg) {
 }
 
 ExpectedKeyValuesMap ParseKeyValues(const vector<string> &items, char delimiter) {
-	KeyValuesMap ret;
-	error::Error err = AddParseKeyValues(ret, items, delimiter);
-	if (err) {
-		return ExpectedKeyValuesMap(err);
-	} else {
-		return ExpectedKeyValuesMap(ret);
-	}
-}
-
-error::Error AddParseKeyValues(KeyValuesMap &base, const vector<string> &items, char delimiter) {
+	KeyValuesMap base;
 	string invalid_data = "";
 
 	for (auto str : items) {
@@ -84,10 +75,9 @@ error::Error AddParseKeyValues(KeyValuesMap &base, const vector<string> &items,
 			"Invalid data given: '" + invalid_data + "'");
 	}
 
-	return MakeError(KeyValueParserErrorCode::NoError, "");
+	return base;
 }
 
-
 } // namespace key_value_parser
 } // namespace common
 } // namespace mender
```

Something like this?

Reply:
discussed f2f (summary):

I _personally_ prefer explictly returning the type, and merging, but merging through side-effects are fine. Do whatever you think works best

Reply:
Could also return a shared_ptr if value-copy is expensive. But modification is also fine, IMHO.